rig/place_and_route/utils.py

```diff
@@ -4,13 +4,187 @@
+                   chip_resources={
+                       core_resource: max_cores,
+                       sdram_resource: max_sdram,
+                       sram_resource: max_sram,
+                   },
+                   chip_resource_exceptions={
+                       chip: {
+                           core_resource: info.num_cores,
+                           sdram_resource: info.largest_free_sdram_block,
+                           sram_resource: info.largest_free_sram_block,
+                       }
+                       for chip, info in iteritems(system_info)
+                       if (info.num_cores != max_cores
+                           or info.largest_free_sdram_block != max_sdram
+                           or info.largest_free_sram_block != max_sram)
+                   },
+                   dead_chips=set(system_info.dead_chips()),
+                   dead_links=set(system_info.dead_links()))
+
+
+def build_core_constraints(system_info, machine, core_resource=Cores):
+    """Return a set of place-and-route
+    :py:class:`~rig.place_and_route.constraints.ReserveResourceConstraint`
+    which reserve any cores that that are already in use.
+
+    The returned list of
+    :py:class:`~rig.place_and_route.constraints.ReserveResourceConstraint`\ s
+    reserves all cores not in an Idle state (i.e. not a monitor and not already
+    running an application).
+
+    .. note::
+
+        Historically it was required that a
+        :py:class:`~rig.place_and_route.constraints.ReserveResourceConstraint`
+        was manually added by every application which reserved the monitor core
+        on each chip. Improves on this by automatically generating appropriate
+        constraints for reserving not just monitor cores but also other cores
+        which are already in use.
+
+    Parameters
+    ----------
+    system_info : :py:class:`rig.machine_control.machine_controller.SystemInfo`
+        The resource availability information for a SpiNNaker machine,
+        typically produced by
+        :py:meth:`rig.machine_control.MachineController.get_system_info`.
+    machine : :py:class:`~rig.machine.Machine`
+        The :py:class:`~rig.machine.Machine` model of the current SpiNNaker
+        machine (e.g. from :py:func:`.build_machine`).
+    core_resource : resource (Default: :py:data:`~rig.machine.Cores`)
+        The resource identifier used for cores.
+
+    Returns
+    -------
+    [:py:class:`rig.place_and_route.constraints.ReserveResourceConstraint`, \
+            ...]
+        A set of place-and-route constraints which reserves all non-idle. The
+        resource type given in the ``core_resource`` argument will be reserved
+        accordingly.
+    """
+    constraints = []
+
+    for x, y in machine:
+        chip_info = system_info[(x, y)]
+
+        # Construct a minimal set of ReserveResourceConstraints which
+        # reserve the cores which are not in the idle state.
```

Comment:
This is certainly true. I was debating what to do about global reservations... You'd have to iterate over the chips twice and do a bitwise-and of the non-idle cores, globally reserve that and then go through everything again reserving all the exceptions.

In terms of the runtime expense of this for the allocation/placement algorithms, generating one global constraint vs many local ones has little or no performance difference. I suppose really I should do global things properly, if nothing else but for human readability...